### PR TITLE
[FEATURE ember-metal-ember-assign] Introduce Ember.assign()

### PR DESCRIPTION
**Pull request by tricknotes**
`Ember.assign()` is discussed in https://github.com/emberjs/ember.js/pull/11978 .
- [x] Add feature flagging for the API addition
- [x] Add a test for the deprecation of `Ember.merge` when the flag is enabled
- [x] Add some cursory tests for `ember-metal/assign` (doesn't need to be complicated, but I noticed no tests were tweaked/changed for `merge` or `assign` and figured we probably want some to exist).
- [x] We should use `Object.assign` if it exists (making our `Ember.assign` a straight up polyfill).

### CODE REVIEW

File: packages/ember-metal/lib/main.js [hunk 1]
```diff
@@ -321,6 +322,7 @@ Ember.isEmpty = isEmpty;
 Ember.isBlank = isBlank;
 Ember.isPresent = isPresent;
 
+Ember.assign = assign;
```
Comment (rwjblue) [inline packages/ember-metal/lib/main.js]: I think we will need to feature flag the public API change.

File: packages/ember-metal/lib/merge.js [hunk 2]
```diff
@@ -16,6 +18,8 @@
   @public
 */
 export default function merge(original, updates) {
+  deprecate('Usage of `Ember.merge` is deprecated, use `Ember.assign` instead.', false, { id: 'ember-metal.merge', until: '3.0.0' });
```
Comment (rwjblue) [inline packages/ember-metal/lib/merge.js]: This will need to be flagged (so that we do not deprecate if the new feature flag is disabled).

**Comment by rwjblue** [comment]
Could you:

* Add feature flagging for the API addition
* Add a test for the deprecation of `Ember.merge` when the flag is enabled
* Add some cursory tests for `ember-metal/assign` (doesn't need to be complicated, but I noticed no tests were tweaked/changed for `merge` or `assign` and figured we probably want some to exist).

**Comment by rwjblue** [comment]
Another point made in #12320 (that I agree with) is that we should use `Object.assign` if it exists (making our `Ember.assign` a straight up polyfill).

**Comment by rwjblue** [comment]
@tricknotes - I moved the various comments/todo items into a checklist in the PR description (so you can  check them off as you go).

**Comment by tricknotes** [comment]
Sorry for my late :<
Thanks @rwjblue . Yes, I'm agree with these todo checklist!

I'll continue to work in this weekend.

**Comment by oriSomething** [comment]
if there is a well known and widely use of a [polyfill](https://www.npmjs.com/package/object-assign), why writing a code for it and not using it?

**Comment by tricknotes** [comment]
> if there is a well known and widely use of a polyfill, why writing a code for it and not using it?

I think `Ember.assign` is enough to work just merge some objects. It might not required to be strict as `Object.assign`.

**Comment by tricknotes** [comment]
@rwjblue 
All todo things are done.
Could you review again?

**Comment by rwjblue** [comment]
@tricknotes - Looks great, can you squash commits and make the new commit's title  `[FEATURE ember-metal-ember-assign]`?

**Comment by tricknotes** [comment]
@rwjblue 
Thanks, @rwjblue .
done.

**Comment by rwjblue** [comment]
Awesome work as usual, thank you @tricknotes!

**Comment by oriSomething** [comment]
@tricknotes @rwjblue i think if `Ember.assign` is supposed to be a polyfill when there is no `Object.assign`, it suppose to support also copying of `Symbol`s as `Object.assign` does.

There are versions of chrome that support `Symbol` but no `Object.assign`, so need is exist.

**Comment by rwjblue** [comment]
We are not intending to make a full fidelity polyfill here, just the behaviors that we need. If you need/want a full polyfill you should include a full fidelity version (like the one you linked) and that will be used instead (since it actually sets Object.assign).

**Comment by oriSomething** [comment]
@rwjblue but three things:
- **it might lead to confusion**. there are still browsers that support `Symbol`s and not `Object.assign` if it's not a polyfill, why naming as one and more over use `Object.assign` when available? better left with `merge` and prevent confusion
- **it will lead to unexpected behavior**, maybe is seems like very edge case because `Symbol`s are not in a wide usage, be it still might cause a bugs and unexpected behaviors in some browsers. specially in future where more people will use them, and more mobile phone with various mobile browsers
- **it's only a few lines of code to add to support `Symbol`s** and make it a full polyfill. much shorter than this comment

i know it sounds picky and very edge case, but as web developers we all fight very edge cases everyday. and many don't know enough `Ember` nor `Object.assign` and we might cause them many trouble for so little code. **in the bottom we better not use it as a polyfill or fully polyfill this feature**

**Pull request merged**